image/codecs/xan.cpp

```diff
@@ -0,0 +1,503 @@
+	if (mode) {
+		// YUV420 frame
+		for (int y = 0; y < _height / 2; y++) {
+			for (int x = 0; x < pitch; x++) {
+				if (src >= src_end)
+					return true;
+				int toff = *src++ * 2;
+				if (toff) {
+					if (toff > table_size)
+						return false;
+					const int pos = stream.pos();
+					stream.seek(table_start + toff);
+					uint8 uval, vval;
+					if (_wc4Mode) {
+						uint16 val = stream.readUint16LE();
+						uval = (val >> 3) & 0xF8;
+						vval = (val >> 8) & 0xF8;
+					} else {
+						uval = stream.readByte();
+						vval = stream.readByte();
+					}
+					uval = uval | uval >> 5;
+					vval = vval | vval >> 5;
+					stream.seek(pos);
+					U[x] = uval;
+					V[x] = vval;
+				}
+			}
+			U += pitch;
+			V += pitch;
+		}
+		if (_height % 1) {
+			memcpy(U, U - pitch, pitch);
+			memcpy(V, V - pitch, pitch);
+		}
+	} else {
+		// YUV410 frame - expand out U and V components
+		uint8 *U2 = U + pitch;
+		uint8 *V2 = V + pitch;
+
+		for (int y = 0; y < _height / 4; y++) {
+			for (int x = 0; x < pitch; x += 2) {
+				if (src >= src_end)
+					return true;
+				int toff = *src++ * 2;
+				if (toff) {
+					if (toff > table_size)
+						return false;
+					const int pos = stream.pos();
+					stream.seek(table_start + toff);
+					uint8 uval, vval;
+					if (_wc4Mode) {
+						uint16 val = stream.readUint16LE();
+						uval = (val >> 3) & 0xF8;
+						vval = (val >> 8) & 0xF8;
+					} else {
+						uval = stream.readByte();
+						vval = stream.readByte();
+					}
+					uval = uval | uval >> 5;
+					vval = vval | vval >> 5;
+					stream.seek(pos);
+					U[x] = U[x+1] = U2[x] = U2[x+1] = uval;
+					V[x] = V[x+1] = V2[x] = V2[x+1] = vval;
+				}
+			}
+			U  += pitch * 2;
+			V  += pitch * 2;
+			U2 += pitch * 2;
+			V2 += pitch * 2;
+		}
+		if (_height % 4) {
+			int lines = ((_height + 1) / 2) - (_height / 4) * 2;
+
+			memcpy(U, U - lines * pitch, lines * pitch);
+			memcpy(V, V - lines * pitch, lines * pitch);
+		}
+	}
+
+	return true;
+}
+
+void XanDecoder::decodeFrameType0(Common::SeekableReadStream &stream) {
+	uint32 chroma_offset = stream.readUint32LE();
+	uint32 refines_offset = stream.readUint32LE();
+	uint32 luma_offset = stream.pos();
+	uint32 nbytes = static_cast<uint32>(stream.size());
+
+	if (chroma_offset > nbytes || refines_offset > nbytes) {
+		error("invalid frame type 0 offsets");
+	}
+
+	if (!decodeChroma(stream, chroma_offset)) {
+		warning("xxan chrome decode failed frame type 0");
+		return;
+	}
+
+	stream.seek(luma_offset);
+	decompressLuma(stream);
+
+	//
+	// Expand out the decompressed luma data. For type 0 frames:
+	// * luma vals are 5-bit diffs, where
+	//   * top row values are diffs on the last value
+	//   * and the remaining rows are deltas on the value above.
+	// * every second pixel in x is linearly interpolated from its horizontal neighbours.
+	// * output values are clipped to 6 bits.
+	// * a set of refinements values can adjust luma of interpolated pixels
+	//
+	uint8 *lumadecomp = _scratchbuf;
```

Comment:
this pointer should be const as far as I can see

Reply:
yep, fixed.